### PR TITLE
[4.0][com_finder] wrong encoding for bytea postgresql

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
wrong encoding for bytea
wrong inner join update syntax

### Testing Instructions
on postgresql
enable smart search plugin
edit & save a tag

### Expected result
tag is saved
works as before on mysql

### Actual result
multiple errors

### CODE REVIEW

**Comment by HLeithner** [comment]
All new pr should use prepared statements, could you update this pr?

**Comment by alikon** [comment]
well yes, i could, but atm with this silly PR i just want to give to the few postgresql users the opportunity to play a little bit with j4, and help us to discover more bugs.....prepared statement conversion should be matter for another pr imho

**Comment by HLeithner** [comment]
At the moment only you and me did prepared statement conversion it should be a strict requirement on all pr touching a sql statement to convert it to use prepared statements.
There was a PLT motion about this topic:

https://volunteers.joomla.org/departments/production/reports/1013-production-dept-meeting-may-07-and-may-14-2019

If we don't required this we will still get new non prepared queries in patches.

**Comment by alikon** [comment]
i'd be more than happy if this motion have been taken in serious count before merging  all & more,
but honestly this has not been happened....

**Comment by mbabker** [comment]
> At the moment only you and me did prepared statement conversion it should be a strict requirement on all pr touching a sql statement to convert it to use prepared statements.

Forcing unrelated changes into pull requests is exactly how you end up making pull requests more complex to review and more error prone.  Converting existing queries should be pull requests of their own, writing new queries it is fine to mandate this.  "You touched it so you need to change it" isn't a good workflow.  Same goes with the random unrelated code style fixes that are commonly requested.

**Comment by HLeithner** [comment]
> Forcing unrelated changes into pull requests is exactly how you end up making pull requests more complex to review and more error prone. Converting existing queries should be pull requests of their own, writing new queries it is fine to mandate this. "You touched it so you need to change it" isn't a good workflow. Same goes with the random unrelated code style fixes that are commonly requested.

This is partly true, but whats the better way? If no queries get updated to PS only if we find more then 2 people to do this job we will never finish this task.

Anyway I don't request to update all queries in this file only the one that get touched and tested afterwards. In a perfect would we would have tests at the same time, or the PR includes a test if non exists but as we all know tests are still missing all over the place. 

It's ok for me if at least new queries use PS...

File: administrator/components/com_finder/Indexer/Driver/Postgresql.php [hunk 1]
```diff
@@ -379,7 +379,7 @@ public function index($item, $format = 'html')
 		$query->clear()
 			->update($db->quoteName('#__finder_links'))
 			->set($db->quoteName('md5sum') . ' = ' . $db->quote($curSig))
-			->set($db->quoteName('object') . ' = ' . $db->quote(serialize($item)))
+			->set($db->quoteName('object') . ' = ' . $db->quote(utf8_encode($item)))
 			->where($db->quoteName('link_id') . ' = ' . $db->quote($linkId));
```
Comment (Hackwar) [inline administrator/components/com_finder/Indexer/Driver/Postgresql.php]: This is not a string. Utf8_encode is wrong here.

File: administrator/components/com_finder/Indexer/Driver/Postgresql.php [hunk 2]
```diff
@@ -132,7 +132,7 @@ public function index($item, $format = 'html')
 		{
 			// Update the link.
 			$entry->link_id = $linkId;
-			$db->updateObject('#__finder_links', $entry, 'link_id');
+			$db->updateObject('#__finder_links', utf8_encode($entry), 'link_id');
 		}
```
Comment (Hackwar) [inline administrator/components/com_finder/Indexer/Driver/Postgresql.php:137]: This is also not correct. The string should already be utf8. If it isn't, there is an issue further up in the code.

**Review by Hackwar — CHANGES_REQUESTED** [review]

**Comment by richard67** [comment]
@alikon Will you work on this, make Harald happy with just changing those queries to PS which you have modified, and handle Hannes‘ comments? If so, ping me when ready for tests. But if you have little time: Shall I help with picking up that work and making PR for your repo so you just can merge with 1 click? If so, ping me and I will do later today.

File: administrator/components/com_finder/Indexer/Driver/Postgresql.php (hunk 2 above)
Comment (alikon) [reply administrator/components/com_finder/Indexer/Driver/Postgresql.php:137]: used pg_escape_bytea instead

File: administrator/components/com_finder/Indexer/Driver/Postgresql.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_finder/Indexer/Driver/Postgresql.php]: used pg_escape_bytea instead

**Review by zero-24 — APPROVED** [review]
I can not acutally check on postgresql but on review it looks good to me.

**Comment by richard67** [comment]
@Hackwar Could you check if it ok now?

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on c484f45da677d3b848053fbe6050f0201da09252<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25884">issues.joomla.org/tracker/joomla-cms/25884</a>.</sub>

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on c484f45da677d3b848053fbe6050f0201da09252<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25884">issues.joomla.org/tracker/joomla-cms/25884</a>.</sub>

**Comment by richard67** [comment]
@Quy RTC please :smile:

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25884">issues.joomla.org/tracker/joomla-cms/25884</a>.</sub>

File: administrator/components/com_finder/Indexer/Indexer.php [hunk 3]
```diff
@@ -287,9 +288,10 @@ public function remove($linkId)
 		// Update the link counts for the terms.
 		$query->clear()
 			->update($db->quoteName('#__finder_terms', 't'))
-			->join('INNER', $db->quoteName('#__finder_links_terms', 'm') . ' ON ' . $db->quoteName('m.term_id') . ' = ' . $db->quoteName('t.term_id'))
+			->join('INNER', $db->quoteName('#__finder_links_terms', 'm'), $db->quoteName('m.term_id') . ' = ' . $db->quoteName('t.term_id'))
 			->set($db->quoteName('links') . ' = ' . $db->quoteName('links') . ' - 1')
-			->where($db->quoteName('m.link_id') . ' = ' . (int) $linkId);
+			->where($db->quoteName('m.link_id') . ' = :linkid')
```
Comment (wilsonge) [inline administrator/components/com_finder/Indexer/Indexer.php:294]: If we're parameterzing this shouldn't we do all the queries using it below this? Also do we not want to keep the `(int) $linkId` part?

Comment (alikon) [reply administrator/components/com_finder/Indexer/Indexer.php:294]: done

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**